### PR TITLE
Refactor ExtVMFace

### PR DESCRIPTION
**Pull request by chfast**
This moves utility structs like `LogEntry` from ExtVMFace.h (libevm) to libethcore. Previously some files from libethcore have to include ExtVMFace.h.

Also `Address` is moved from libdevcrypto to libdevcore as it is used not only in crypto context.

### CODE REVIEW

**Comment by winsvega** [comment]
@pirapira

File: libdevcore/FixedHash.cpp [hunk 1]
```diff
@@ -14,20 +14,25 @@
...
 #include <boost/algorithm/string.hpp>
 
 using namespace std;
 using namespace dev;
 
-std::random_device dev::s_fixedHashEngine;
+namespace dev
+{
+
+const Address ZeroAddress = Address();
```
Comment (gumb0) [inline libdevcore/FixedHash.cpp]: `Address const`

File: libdevcore/FixedHash.cpp [hunk 2]
```diff
@@ -14,20 +14,25 @@
...
-/** @file FixedHash.cpp
- * @author Gav Wood <i@gavwood.com>
- * @date 2014
- */
 
 #include "FixedHash.h"
 #include <boost/algorithm/string.hpp>
 
 using namespace std;
 using namespace dev;
```
Comment (gumb0) [inline libdevcore/FixedHash.cpp]: I think we can remove this now

File: libdevcore/FixedHash.h [hunk 3]
```diff
@@ -351,6 +351,25 @@ using h160Set = std::set<h160>;
 using h256Hash = std::unordered_set<h256>;
 using h160Hash = std::unordered_set<h160>;
 
+/// An Ethereum address: 20 bytes.
+/// @NOTE This is not endian-specific; it's just a bunch of bytes.
+using Address = h160;
```
Comment (gumb0) [inline libdevcore/FixedHash.h]: I'd prefer to put these into separate Address.h/cpp

File: libethcore/Transaction.h [hunk 4]
```diff
@@ -14,19 +14,16 @@
...
 	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
-/** @file TransactionBase.h
- * @author Gav Wood <i@gavwood.com>
- * @date 2014
- */
 
 #pragma once
 
+#include <boost/optional.hpp>
```
Comment (gumb0) [inline libethcore/Transaction.h]: `#include`s should be in increasing order of generality, so this goes after our own headers

File: libethereum/TransactionReceipt.h [hunk 5]
```diff
@@ -22,11 +22,9 @@
 #pragma once
 
 #include <array>
-#include <libdevcore/Common.h>
-#include <libdevcore/RLP.h>
-#include <libevm/ExtVMFace.h>
-
 #include <boost/variant/variant.hpp>
+#include <libdevcore/RLP.h>
```
Comment (gumb0) [inline libethereum/TransactionReceipt.h]: should be before boost & `<array>`

File: libethcore/Common.h [hunk 6]
```diff
@@ -90,6 +94,81 @@ enum class RelativeBlock: BlockNumber
...
 };
 
+enum class BlockPolarity
+{
+	Unknown,
+	Dead,
+	Live
+};
+
+struct LogEntry
```
Comment (gumb0) [inline libethcore/Common.h]: These could go into separate `LogEntry.h`, but it's up to you

File: libethcore/Transaction.h (hunk 4 above)
Comment (chfast) [reply libethcore/Transaction.h]: I think we have it the other way around in most files. Personally, I don't care. Why is this better?

File: libdevcore/FixedHash.h (hunk 3 above)
Comment (chfast) [reply libdevcore/FixedHash.h]: Two cons for this: increased build time (?) and additional maintenance cost. I'm not sure...

File: libethcore/Transaction.h (hunk 4 above)
Comment (gumb0) [reply libethcore/Transaction.h]: Link from coding standards: http://stackoverflow.com/questions/614302/c-header-order/614333#614333

It prevents the headers from not including all necessary headers but still working because of weird inter-header dependency

File: libdevcore/FixedHash.h (hunk 3 above)
Comment (gumb0) [reply libdevcore/FixedHash.h]: I think it's your way it could be theoretically increased build time, because everyone who needs only `FixedHash` has to compile `Address`-related stuff, too

Comment (gumb0) [reply libdevcore/FixedHash.h]: It would make navigating the code easier, putting `Address` into Address.h instead of suprisingly `FixedHash.h`

File: libethcore/BasicAuthority.h [hunk 7]
```diff
@@ -23,6 +23,7 @@
 
 #pragma once
 
+#include <libdevcrypto/Common.h>
```
Comment (pirapira) [inline libethcore/BasicAuthority.h]: Why is this added now?

File: libethcore/Common.cpp [hunk 8]
```diff
@@ -191,5 +189,26 @@ string TransactionSkeleton::userReadable(bool _toProxy, function<pair<bool, stri
 		formatBalance(gas * gasPrice) + ".");
 }
 
+LogEntry::LogEntry(RLP const& _r)
+{
```
Comment (pirapira) [inline libethcore/Common.cpp]: Should there be an assertion against
- `_r` not being a list and
- `_r` being too short?

File: libethcore/Common.cpp [hunk 9]
```diff
@@ -191,5 +189,26 @@ string TransactionSkeleton::userReadable(bool _toProxy, function<pair<bool, stri
...
+	address = (Address)_r[0];
```
Comment (pirapira) [inline libethcore/Common.cpp]: Rename it to `m_address`.

Comment (pirapira) [reply libethcore/Common.cpp]: Ahh, it's a struct. Nevermind.

File: libethcore/Common.h [hunk 10]
```diff
@@ -90,6 +94,81 @@ enum class RelativeBlock: BlockNumber
...
+{
+	LogEntry() = default;
+	explicit LogEntry(RLP const& _r);
+	LogEntry(Address const& _address, h256s _ts, bytes&& _d): address(_address), topics(std::move(_ts)), data(std::move(_d)) {}
```
Comment (pirapira) [inline libethcore/Common.h]: I wonder why `_ts` is not an rvalue reference and `_d` is an rvalue reference...

Comment (chfast) [reply libethcore/Common.h]: Generic way is to pass by value and move in the initialization list (like with `h256s` case). This way you handle both copy and move, depending what argument you have.

The version with `&&` prevents you from passing argument by copy -- you can only pass an argument being rvalue. This is restrictive, but enforces some optimizations. I don't like it because it is more confusing than useful.

File: libethcore/BasicAuthority.h (hunk 7 above)
Comment (chfast) [reply libethcore/BasicAuthority.h]: Previously this header was implicitly included for Address. Now we still need it for crypto::Signature.

**Comment by chfast** [comment]
I think I addressed all issues. Can you review before builds?

File: libethcore/Common.h [hunk 11]
```diff
@@ -90,6 +96,13 @@ enum class RelativeBlock: BlockNumber
 	Pending = PendingBlock
 };
 
+enum class BlockPolarity
```
Comment (gumb0) [inline libethcore/Common.h]: This is used only in `LocalisedLogEntry`, maybe move it to `LogEntry.h`?

Comment (chfast) [reply libethcore/Common.h]: Unfortunately, this is also used in Client.cpp and ClientBase.cpp.
